homeassistant/components/swidget/config_flow.py

```diff
@@ -0,0 +1,211 @@
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    device: SwidgetDevice = SwidgetDevice(
+        host=data["host"],
+        token_name="x-secret-key",
+        secret_key=data.get("password"),
+        use_https=True,
+        use_websockets=False,
+    )
+    try:
+        await device.update()
+        await device.stop()
+    except SwidgetException as exc:
+        raise CannotConnect from exc
+    return {
+        "title": device.friendly_name,
+        "mac_address": format_mac(device.mac_address),
+    }
+
+
+class SwidgetConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for swidget."""
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._discovered_devices: dict[str, SwidgetDiscoveredDevice] = {}
+        self._discovered_device: SwidgetDiscoveredDevice | None = None
+
+    async def async_step_dhcp(
+        self, discovery_info: dhcp.DhcpServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle discovery via dhcp."""
+        _LOGGER.debug("Swidget device found via DHCP: %s", discovery_info)
+        return await self._async_handle_discovery(
+            discovery_info.ip, format_mac(discovery_info.macaddress)
+        )
+
+    async def async_step_ssdp(
+        self, discovery_info: ssdp.SsdpServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle discovery via SSDP."""
+        _LOGGER.debug("Swidget device found via SSDP: %s", discovery_info)
+        discovered_ip = urlparse(discovery_info.ssdp_headers["location"]).hostname
+        discovered_mac = format_mac(discovery_info.ssdp_headers["USN"].split("-")[-1])
+        device_type = discovery_info.ssdp_headers["SERVER"].split(" ")[1].split("+")[0]
+        insert_type = (
+            discovery_info.ssdp_headers["SERVER"]
+            .split(" ")[1]
+            .split("+")[1]
+            .split("/")[0]
+        )
+        friendly_name = discovery_info.ssdp_headers["SERVER"].split("/")[2].strip('"')
+        return await self._async_handle_discovery(
+            host=discovered_ip,
+            mac=discovered_mac,
+            device_type=device_type,
+            insert_type=insert_type,
+            friendly_name=friendly_name,
+        )
+
+    async def async_step_integration_discovery(
+        self, discovery_info: DiscoveryInfoType
+    ) -> ConfigFlowResult:
+        """Handle integration discovery."""
+        return await self._async_handle_discovery(
+            discovery_info[CONF_HOST], discovery_info[CONF_MAC]
+        )
+
+    async def _async_handle_discovery(
+        self,
+        host: str,
+        mac: str,
+        device_type: str = "Unknown",
+        insert_type: str = "Unknown",
+        friendly_name: str = "Unknown Swidget Device",
+    ) -> ConfigFlowResult:
+        """Handle any discovery."""
+        await self.async_set_unique_id(format_mac(mac))
+        self._abort_if_unique_id_configured(updates={CONF_HOST: host})
```

Comment:
This means we'll automatically update the IP-address, which is nice. However, I think you also need to create a new `SwidgetDevice` or tell it about the new IP-address?

The easiest pattern is to reload the config entry:

```
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
   ...
    entry.async_on_unload(entry.add_update_listener(config_entry_update_listener))
    return True


async def config_entry_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Update listener, called when the config entry options are changed."""
    await hass.config_entries.async_reload(entry.entry_id)
```

Reply:
Updated. Please check

Reply:
What's triggering the update listener?